File: interview_prep/asari/ringbuffer.py
#!/usr/bin/env python

class RingBuffer:
    def __init__(self, capacity):
        self.buffer = [None] * capacity
        self.capacity = capacity
        self.head = 0
        self.tail = 0
        self.full = False

    def push(self, item):
        self.buffer[self.tail] = item
        if self.full:
            self.head = (self.head + 1) % self.capacity
        self.tail = (self.tail + 1) % self.capacity
        self.full = self.tail == self.head

    def pop(self):
        if self.empty():
            raise IndexError("Buffer is empty")
        item = self.buffer[self.head]
        self.full = False
        self.head = (self.head + 1) % self.capacity
        return item

    def empty(self):
        return (not self.full) and (self.head == self.tail)

    def is_full(self):
        return self.full

    def size(self):
        if self.full:
            return self.capacity
        if self.tail >= self.head:
            return self.tail - self.head
        return self.capacity - self.head + self.tail

File: interview_prep/asari/ring_buffer.cpp
#include <iostream>
#include <vector>

template<typename T>
class RingBuffer {
public:
    RingBuffer(size_t capacity)
        : buffer(capacity), head(0), tail(0), full(false) {}

    void push(const T& item) {
        buffer[tail] = item;
        if (full) {
            head = (head + 1) % buffer.size();
        }
        tail = (tail + 1) % buffer.size();
        full = tail == head;
    }

    T pop() {
        if (empty()) {
            throw std::runtime_error("Buffer is empty");
        }
        T item = buffer[head];
        full = false;
        head = (head + 1) % buffer.size();
        return item;
    }

    bool empty() const {
        return (!full && (head == tail));
    }

    bool is_full() const {
        return full;
    }

    size_t size() const {
        if (full) {
            return buffer.size();
        }
        if (tail >= head) {
            return tail - head;
        }
        return buffer.size() - head + tail;
    }

private:
    std::vector<T> buffer;
    size_t head;
    size_t tail;
    bool full;
};

// Example usage
int main() {
    RingBuffer<int> rb(3);
    rb.push(1);
    rb.push(2);
    rb.push(3);
    std::cout << rb.pop() << std::endl; // 1
    rb.push(4);
    std::cout << rb.pop() << std::endl; // 2
    std::cout << rb.pop() << std::endl; // 3
    std::cout << rb.pop() << std::endl; // 4
    return 0;
}
